numpy/ctypeslib.py

```diff
@@ -287,36 +291,48 @@ def ndpointer(dtype=None, ndim=None, shape=None, flags=None):
             except Exception:
                 raise TypeError("invalid flags specification")
             num = _num_fromflags(flags)
+    
+    # Compose cache key tuple:
+    cache_key = (dtype, ndim, shape, num)
+    
     try:
-        return _pointer_type_cache[(dtype, ndim, shape, num)]
+        return _pointer_type_cache[cache_key]
     except KeyError:
         pass
+    
     if dtype is None:
         name = 'any'
     elif dtype.names:
         name = str(id(dtype))
     else:
         name = dtype.str
+    
     if ndim is not None:
         name += "_%dd" % ndim
+    
     if shape is not None:
         try:
             strshape = [str(x) for x in shape]
         except TypeError:
             strshape = [str(shape)]
             shape = (shape,)
```

Comment:
This line should happen before the cache lookup

Reply:
Oh OK I just saw this message – so just to be sure, all the normalization and name-munging should happen _before_ composing the cache key, yes?

Reply:
The name-munging should happen after, but the normalization should happen before

Reply:
OK right – I would assume that a double call to `type(…)` with the same args results in some kind of problematic disruption of the runtime class tree, in that case.